### ISSUE TITLE
CMake fixes for FLTK 1.4.0

### ISSUE THREAD

**Issue by Albrecht-S** [fltk/fltk#655]
This issue is my personal reminder of necessary fixes related to CMake. My plan is to extend the task list in this comment.

Please do not add other CMake related comments to this issue, I'll maintain this issue until 1.4.0 gets released. It will be incremented slowly whenever I find a (new) CMake issue.

The following points are currently maintained in STR's which I will close for easier maintenance:
- [x] [STR 3471](https://www.fltk.org/str.php?L3471): "CMake improvements": **closed**, **fixed in 63c54fe7c8eda4652ef81ab9a759e181ac48c5ee**
- [x] [STR 3224](https://www.fltk.org/str.php?L3224): "add CMake package registry option": **closed**, **won't fix.** See [comment below](https://github.com/fltk/fltk/issues/655#issuecomment-1858827748).
- [x] [STR 3443](https://www.fltk.org/str.php?L3443): "cmake/Visual Studio: control /MT vs /MD compiler flags": **closed**, fixed in commit 4a461efae6fa36aeb50e61e381dc2ecf95b83d23 (1.4.0 Git) and in release 1.3.9 as well.

The following points should be evaluated but are NOT showstoppers for release 1.4.0:
- [x] [STR 3505](https://www.fltk.org/str.php?L3505): "Add FLTK as package for the Hunter packaging system": ~**open**, not sure how to fix this.~ **_Edit:_** closed, will hopefully be fixed by current work on CMake. If anything doesn't work out we need a new GitHub Issue.
- [x] [STR 3415](https://www.fltk.org/str.php?L3415): "Programs may need some definitions from FLTK to work correctly": closed, fixed by move to "modern CMake", see also comment on STR 3415.
- [x] [STR 3264](https://www.fltk.org/str.php?L3264): "The autotools build system does not build FLTKConfig.cmake": ~**open**, not sure how to fix this.~ **_Edit:_** closed, won't fix in 1.4.x and FLTK 1.5 will no longer support autotools.

Other CMake issues:
- [x] Add `FLTKConfigVersion.cmake` to provide consumers with FLTK version info. See https://cmake.org/cmake/help/latest/module/CMakePackageConfigHelpers.html
- [x] Maybe add support for "Components" - low priority, not in 1.4.0 but maybe later. 
- [x] Improve OpenGL searching, error messages, and related stuff, see e.g. issue #1037. Not in 1.4.0, will be done later.

**Comment by Albrecht-S** [comment]
Loosely related [STR 2642](https://www.fltk.org/str.php?L2642): _Need examples and docs for how to build apps against DLL version of FLTK_.

This STR is not entirely `CMake` related but the documentation would be (mostly) based on `CMake` techniques.

**Comment by Albrecht-S** [comment]
It's documented in README.CMake.txt (see below) but I must admit that I didn't remove or rewrite your advice yet, sorry for the confusion. As you noticed, it fails to do what you did before. The correct way is now to use CMake's commandline switch "`-D FLTK_MSVC_RUNTIME_DLL=OFF`" to achieve what you want:
```
FLTK_MSVC_RUNTIME_DLL - default ON (only for Visual Studio and NMake).
   Selects whether the build uses the MS runtime DLL or not.
   Default is ON: either /MD or /MDd for Release or Debug, respectively.
   Select OFF for either /MT or /MTd for Release or Debug, respectively.
```
Note that this works (and is documented) in FLTK 1.3.9 as well if and only if you use CMake 3.15 or later (in 1.4 this is the minimum CMake version allowed, but in 1.3.9 you can still use older CMake versions).

The reason why I didn't change the NMake instructions yet was that I  wanted to test it myself. I tried it yesterday successfully but I still need to update the docs. For best compatibility you should set neither `CMAKE_CXX_FLAGS_*` nor `CMAKE_C_FLAGS_*` on the commandline.

Side note: I also noticed that CMake 3.15 and later does not set default warning levels /W3 or /W4 or whatever. This applies only to FLTK 1.4 though and I'm still looking for a good solution and will document this as well.

**Comment by Albrecht-S** [comment]
> So it sounds like my /MT flag in the cmake is fighting against /MD (which I don't want), yet /MD is somehow "winning".

I guess your setup included both options (your `/MT` before CMake's `-MD`) and the latter overwrites the former - or maybe not, I don't really know.

You also wrote that you used `-D CMAKE_MSVC_RUNTIME_LIBRARY=MultiThreaded` which **seems** to be (almost) correct (CMake uses much more complicated "generator expressions" to define it for all build types in the IDE as also noted in the CMake docs). However, I believe that our new option overwrites your commandline switch internally. Maybe I could check if this specific CMake variable has been set by the user and then skip the internal setting, but this would make it even more difficult (a) to document the behavior and (b) for the user. Please try the new option described in README.CMake.txt **_w/o setting_** `CMAKE_MSVC_RUNTIME_LIBRARY`. This should always be correct and is much simpler for the user (you) which is why I defined this option in the first place.

**Comment by Albrecht-S** [comment]
> according to the build files, which is showing:
> CXX_FLAGS = /DWIN32 /D_WINDOWS /GR /EHsc /O2 /Ob2 /DNDEBUG -MD /utf-8

Which "build files" are showing this? I was looking around for the full commandlines but CMake + NMake hide this pretty well. Can you help me?

**Comment by erco77** [comment]
>>    CXX_FLAGS = /DWIN32 /D_WINDOWS /GR /EHsc /O2 /Ob2 /DNDEBUG -MD /utf-8
>
> Which "build files" are showing this? I was looking around for the full commandlines but CMake +
> NMake hide this pretty well. Can you help me?

Yeah, the world of Windows + Visual Studio is horrible; who knows where important things like logs and such go.

I had to go grepping around, and found these files in the build directory:

    C:\temp\fltk-1.4.x.git\build\test\CMakeFiles\threads.dir\flags.make
    C:\temp\fltk-1.4.x.git\build\test\CMakeFiles\tile.dir\flags.make
    C:\temp\fltk-1.4.x.git\build\test\CMakeFiles\tiled_image.dir\flags.make
    C:\temp\fltk-1.4.x.git\build\test\CMakeFiles\tree.dir\flags.make

..and if you look at one, the files are very short; you can see the definitions of the CXX_* flags, which in my case look like:

```
# CMAKE generated file: DO NOT EDIT!
# Generated by "NMake Makefiles" Generator, CMake Version 3.20

# compile CXX with C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe

CXX_DEFINES = -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_WARNINGS

CXX_INCLUDES = -IC:\temp\fltk-1.4.x.git\build -IC:\temp\fltk-1.4.x.git -IC:\temp\fltk-1.4.x.git\zlib -IC:\temp\fltk-1.4.x.git\jpeg -IC:\temp\fltk-1.4.x.git\png -IC:\temp\fltk-1.4.x.git\test -IC:\temp\fltk-1.4.x.git\build\test

CXX_FLAGS = /DWIN32 /D_WINDOWS /GR /EHsc /O2 /Ob2 /DNDEBUG -MD /utf-8
```

I'm not sure how to work around this or fix it. The only thing I see that says this might not work is in the [cmake docs for CMAKE_MSVC_RUNTIME_LIBRARY](https://cmake.org/cmake/help/latest/variable/CMAKE_MSVC_RUNTIME_LIBRARY.html#variable:CMAKE_MSVC_RUNTIME_LIBRARY) where, down at the bottom it says something that I don't really understand, but sounds like it might be the problem:
_______________________________________________________________________
Note

This variable has effect only when policy [CMP0091](https://cmake.org/cmake/help/latest/policy/CMP0091.html#policy:CMP0091) is set to NEW prior to the first [project()](https://cmake.org/cmake/help/latest/command/project.html#command:project) or [enable_language()](https://cmake.org/cmake/help/latest/command/enable_language.html#command:enable_language) command that enables a language using a compiler targeting the MSVC ABI.
_______________________________________________________________________
I'm not sure what a CMP0091 is, or how to "set to NEW prior to the first project()", but I'm guessing that's worth a try.. I just don't know what it means.

**Comment by erco77** [comment]
I'm tempted to make a script that just substitutes all those -MD flags with -MT flags just to try to get on with life, but even if that works, I'd feel kinda dirty doing it.

**Comment by erco77** [comment]
I wrote the script to replace the -MD flags with -MT, putting it here for what it's worth; works with msys64's version of python:
```
# fix-mt.py - run this from the fltk directory, assumes ./build is the build directory

import os,sys

for root,dirnames,filenames in os.walk("/c/temp/fltk-1.4.x.git/build"):
    for filename in filenames:
        if ( filename == "flags.make" ):     # find all the flags.make files
            path = os.path.join(root,filename)
            sys.stderr.write("WORKING ON %s\n" % path)  # show what we're working on
            os.system("sed -i /-MD/s//-MT/g %s" % path)   # do the deed
```

From what I can tell, this "cmake/nmake files" way of building doesn't generate "log files" per se, the compile/link lines are hidden, not shown in logs or on stdout/err.

From what I can tell one can use `nmake VERBOSE=1` to see some kind of build output, but you'll regret you did; very noisy, and I actually don't think you can even access the CL and LINK commands that way.

I seemed to have better luck setting the cmake flag CMAKE_VERBOSE_MAKEFILE to "on", and I was able to see at least the bare CL commands -- this seemed to work from within a VS 2019 DOS shell:

    mkdir build
    cd build
    cmake -G "NMake Makefiles" -D CMAKE_BUILD_TYPE=Release -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON ..

Then in a separate msys64 window I ran "python fix-mt.py", followed by running 'cd build' and 'nmake > output.log 2>&1' back in the VS 2019 DOS window.

That still outputs a lot of stuff, but the actual CL command lines are at least buried in there, which is good, e.g.

    cl   /TP -DWIN32_LEAN_AND_MEAN -D_CRT_SECURE_NO_WARNINGS -IC:\temp\fltk-1.4.x.git\build -IC:\temp\fltk-1.4.x.git -IC:\temp\fltk-1.4.x.git\zlib -IC:\temp\fltk-1.4.x.git\jpeg -IC:\temp\fltk-1.4.x.git\png -IC:\temp\fltk-1.4.x.git\test -IC:\temp\fltk-1.4.x.git\build\test /DWIN32 /D_WINDOWS /GR /EHsc /O2 /Ob2 /DNDEBUG -MT /utf-8 /showIncludes /FoCMakeFiles\cairo_test.dir\cairo_test.cxx.obj /FdCMakeFiles\cairo_test.dir\ /FS -c C:\temp\fltk-1.4.x.git\test\cairo_test.cxx

..and for linking, well, I'm not sure it gets better than this actually:

       "C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -E vs_link_exe --intdir=CMakeFiles\cairo_test.dir --rc=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\rc.exe --mt=C:\PROGRA~2\WI3CF2~1\10\bin\100220~1.0\x64\mt.exe --manifests -- C:\PROGRA~2\MICROS~2\2019\COMMUN~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\link.exe  @CMakeFiles\cairo_test.dir\objects1.rsp @C:\Users\erco\AppData\Local\Temp\nm1061.tmp

That seems to still be a cmake command wrapper, not an actual link command. That might be a link command at the end, but it seems to refer to a tmp file for the command line; link commands tend to be long because of the list of libs can be long, so Microsoft often puts them in a separate temp file and refer to that file on the command line.

"Where are these linker temp files?" you might ask, so they can be inspected.

Well in my case, apparently it writes them to my home directory; from the above "link command" one can see it's referring to files like `C:\Users\erco\AppData\Local\Temp\nm1061.tmp`, which implies build files are being generated /outside/ the build directory (bad), but I think they're removed immediately anyway (which is also bad, as this prevents reviewing their contents, sadly), so I'm not sure how that's supposed to work.

Doing anything with windows always wants me to take a long shower afterwards, but no amount of scrubbing washes away the pain, lol.

**Comment by erco77** [comment]
> Can you tell whether /MD and -MD are equivalent? 

I'm pretty sure it is. The / vs - thing is mostly a very old DOS thing, and I think even Microsoft admits that was a bad idea, as win32 and the NT kernel does understand front and backslashes equally for paths, so I assume -I/some/path works as well as /I\some\path.

**Comment by erco77** [comment]
> Please try the new option described in README.CMake.txt w/o setting CMAKE_MSVC_RUNTIME_LIBRARY. This should always be correct and is much simpler for the user (you) which is why I defined this option in the first place.

Is this referring to the OPTION_USE_THREADS flag? Sure, I'll give it a try now.

**Comment by erco77** [comment]
Hmm, OK, I cleared and recreated my build directory and tried:

    cmake -G "NMake Makefiles" -D CMAKE_BUILD_TYPE=Release FLTK_MSVC_RUNTIME_DLL=off ..

..I'm still getting -MD flags in the flags.make files:
```
C:\temp\fltk-1.4.x.git\build>more C:\temp\fltk-1.4.x.git\build\test\CMakeFiles\tiled_image.dir\flags.make
[..]
CXX_FLAGS = /DWIN32 /D_WINDOWS /GR /EHsc /O2 /Ob2 /DNDEBUG -MD /utf-8
```
Am I missing something?

**Comment by erco77** [comment]
Oh duh, yeah, I was literally missing something.. the -D flag.

Using the following, now it works.. wonderful!

    cmake -G "NMake Makefiles" -D CMAKE_BUILD_TYPE=Release -D FLTK_MSVC_RUNTIME_DLL=off ..

I'll adjust my entry in the README.CMake.txt file.

**Comment by Albrecht-S** [comment]
@erco77 Greg, thanks for your clarifications (/MD vs -MD and the location of CXX flags etc.), your tests with NMake, and for updating README.CMake.txt. I take it that everything works well for you now.

Since I also added the new CMake option `FLTK_MSVC_RUNTIME_DLL` to [release 1.3.9](https://www.fltk.org/articles.php?L1898) I copied your advice to `README.CMake.txt` [in 1.3.9](https://github.com/fltk/fltk/blob/a66273e72621887fcbd0ee5dca2dd133356d5d8a/README.CMake.txt#L298) as well.

**Comment by Albrecht-S** [comment]
Regarding `CMake` and `CMPxxxx`: CMP stands for "CMake policy" and is used mainly for backwards compatibility of CMake project files so developers and project users can use different CMake versions. Each policy can be set to `OLD` or `NEW` which is explained in the policy's documentation. I don't want to go into details here but the key point is that setting a particular CMake minimum version in CMake file, for instance `cmake_minimum_required(VERSION 3.15)` which we are currently using, initializes all CMake policies that have been added **_in 3.15 or earlier_** to `NEW` and **_all newer policies_** to `OLD`. Basically this means that newer `CMake` versions than 3.15 behave as if they were CMake 3.15. CMake project developers can test the CMake version that is being executed and adjust other policies as they like, and so on. However, if they allow older versions than some (newer) CMake code requires, they must test the CMake version and add fallback code for older versions which is error prone. That's why I raised the minimum CMake version to 3.15 recently - and I will maybe even go higher in the near future.

I hope that makes the CMake docs a little more understandable.

**Comment by Albrecht-S** [comment]
> Doing anything with windows always wants me to take a long shower afterwards, but no amount of scrubbing washes away the pain, lol.

I never tried taking a shower after using Windows but I feel the same. ;-)

I always try to minimize working with Windows and I'm fortunately only using it for FLTK development. Everything else I need can be done with real OS's (Linux, Unix, macOS, or (Open)VMS which I used earlier for work).

**Comment by erco77** [comment]
> Regarding CMake and CMPxxxx

Ah, that's an interesting way of handing backwards compatibility, kinda sorta similar to our ABI version stuff.

**Comment by Albrecht-S** [comment]
Regarding [STR 3224](https://www.fltk.org/str.php?L3224): "add CMake package registry option": **_closed, won't fix_**: see comment on STR 3224 for further info why this is not implemented in FLTK 1.4.0.

A potential implementation would likely be different and use two CMake variables which can be set by the user:
- [CMAKE_EXPORT_PACKAGE_REGISTRY](https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_PACKAGE_REGISTRY.html#variable:CMAKE_EXPORT_PACKAGE_REGISTRY) would enable this feature, i.e. FLTK would use `export(PACKAGE ...)` to write to the package registry.
- a new option or variable (maybe `FLTK_REGISTRY_NAME`) with default value "FLTK" could be used to specify the package registry name, e.g. `FLTK-Debug`, `FLTK-Release`, ...

However, these are only thoughts and there is currently no intention to implement this (unless users ask for it).

**Comment by Albrecht-S** [comment]
All items in this issue have been fixed or will not be fixed in 1.4.0, hence I'm closing this issue now...
